packages/sdk38/src/wallet.ts

```diff
@@ -0,0 +1,131 @@
+}
+
+export interface OfflineSigner {
+  /**
+   * Get AccountData array from wallet. Rejects if not enabled.
+   */
+  readonly getAccounts: () => Promise<readonly AccountData[]>;
+
+  /**
+   * Request signature from whichever key corresponds to provided bech32-encoded address. Rejects if not enabled.
+   */
+  readonly sign: (address: string, message: Uint8Array, prehashType?: PrehashType) => Promise<StdSignature>;
+}
+
+function prehash(bytes: Uint8Array, type: PrehashType): Uint8Array {
+  switch (type) {
+    case null:
+      return new Uint8Array([...bytes]);
+    case "sha256":
+      return new Sha256(bytes).digest();
+    case "sha512":
+      return new Sha512(bytes).digest();
+    default:
+      throw new Error("Unknown prehash type");
+  }
+}
+
+/**
+ * The Cosmoshub derivation path in the form `m/44'/118'/0'/0/a`
+ * with 0-based account index `a`.
+ */
+export function makeCosmoshubPath(a: number): readonly Slip10RawIndex[] {
+  return [
+    Slip10RawIndex.hardened(44),
+    Slip10RawIndex.hardened(118),
+    Slip10RawIndex.hardened(0),
+    Slip10RawIndex.normal(0),
+    Slip10RawIndex.normal(a),
+  ];
+}
+
+export class Secp256k1OfflineWallet implements OfflineSigner {
+  public static async fromMnemonic(
+    mnemonic: string,
+    hdPath: readonly Slip10RawIndex[] = makeCosmoshubPath(0),
+    prefix = "cosmos",
+  ): Promise<Secp256k1OfflineWallet> {
+    const seed = await Bip39.mnemonicToSeed(new EnglishMnemonic(mnemonic));
+    const { privkey } = Slip10.derivePath(Slip10Curve.Secp256k1, seed, hdPath);
+    const uncompressed = (await Secp256k1.makeKeypair(privkey)).pubkey;
+    return new Secp256k1OfflineWallet(privkey, Secp256k1.compressPubkey(uncompressed), prefix);
+  }
+
+  private readonly pubkey: Uint8Array;
+  private readonly privkey: Uint8Array;
+  private readonly prefix: string;
+  private readonly algo: Algo = "secp256k1";
+  private enabled = false;
+
+  private constructor(privkey: Uint8Array, pubkey: Uint8Array, prefix: string) {
+    this.privkey = privkey;
+    this.pubkey = pubkey;
+    this.prefix = prefix;
+  }
+
+  private get address(): string {
+    return rawSecp256k1PubkeyToAddress(this.pubkey, this.prefix);
+  }
+
+  /**
+   * Request access to the user's accounts. Some wallets will ask the user to approve or deny access. Returns true if granted access or false if denied.
+   */
+  private async enable(): Promise<boolean> {
+    this.enabled = true;
+    return this.enabled;
+  }
+
+  public async getAccounts(): Promise<readonly AccountData[]> {
+    if (!this.enabled && !(await this.enable())) {
+      throw new Error("Wallet not enabled");
+    }
+    return [
+      {
+        address: this.address,
+        algo: this.algo,
+        pubkey: this.pubkey,
+      },
+    ];
+  }
+
+  public async sign(
+    address: string,
+    message: Uint8Array,
+    prehashType: PrehashType = "sha256",
+  ): Promise<StdSignature> {
+    if (!this.enabled && !(await this.enable())) {
```

Comment:
The wallet enabling itself is not so useful, right? If we don't have it in the interface anymore, I think we can remove `this.enabled` completely.

Reply:
Yeah, I guess I was trying to showcase how the flow of a more user-interactive wallet might look, but it's not necessary for this simple one.

Reply:
Or I guess somebody could extend this class and override the `enable` method.